Networking v2

j0sh opened the pull request:
Okay, this should be good to review now. Here are some notes:

The `doc/networking.md` API spec can be used to guide the networking part of the review.

https://github.com/livepeer/go-livepeer/pull/447/commits/16ee37889838651507f8cda2a56b2461a989d9f9 : The transcoder flow is hopefully straightforward. The behavior after retrieving a segment and checking the broadcaster signature is roughly this:

* Retrieves a (global) segment channel for the job.
* If this is a new segment channel (max queue depth 1), begin a transcode loop.
* Submit the segment to the channel.
* The transcode loop picks up the segment.
    * Max queue depth of 1 ensures per-job transcodes are sequential (we can handle concurrent jobs)
    * If the segment channel blocked due to an in-progress transcode, this is a signal that something is wrong, either the transcoder is too slow or the broadcaster is being inconsistent (we can check the broadcaster case separately, but we don't right now)
    * I've observed out-of-order segments in the resulting manifest occasionally during testing; needs investigation. Most likely two (quite out of sync) segment requests are coming in simultaneously, and there's a race to get to the segment channel.
* If there are no new segments for 10 minutes, quit the loop.
  * Close the segment channel
  * Call the claimmanager's `ClaimVerifyAndDistributeFees`

Note that the transcode loop duplicates much of the initialization logic in `TranscodeAndBroadcast` in order to avoid making too many changes to existing code.

https://github.com/livepeer/go-livepeer/pull/447/commits/60c7a724e3613b970af8e3fd29d701097e6a4bb1 : Modifies existing API, but hopefully self explanatory: allows for transcoder error feedback. Overall this part is still raw, but a big step up from what we have now.

https://github.com/livepeer/go-livepeer/pull/447/commits/3208d2c456d35bc4848e7275da1ec5aaa8089d06 : Here, the breaking change to the existing flow is to avoid calling `TranscodeSub` via `TranscodeAndBroadcast` . To achieve this, we bypass the transcoder kick-off procedure in the JobService. Note the duplicated claim logic (again, to avoid changing too much existing code); a later cleanup of the codebase around JobService and TranscodeAndBroadcast can remedy this.

https://github.com/livepeer/go-livepeer/pull/447/commits/649d8eaea7c754a745510efadd163418e48c0649 : A minor but useful change to existing API: we use go-ethereum's new Watch API to listen for jobs on the broadcaster, and return the new job in `CreateTranscodeJob` . The caller (within mediaserver) uses the job to start the transcoding process in https://github.com/livepeer/go-livepeer/pull/447/commits/edc531e7703253c3acd38b6f4848983a0b3eeeeb.

https://github.com/livepeer/go-livepeer/pull/447/commits/9ce6e99c8c9d999b5fd0420980d809a1bb363759 : Checking deposit prior to transcoding. I haven't actually been able to test this particular flow yet, so it's a blind change.

https://github.com/livepeer/go-livepeer/pull/447/commits/601bfb8c3f5f81b7b02da81ec9927a9d344e804c : A small discretionary change to the API for `eth.AssignedTranscoder. In all cases this function is used, we already retrieve the job outside the function, so pass in the job instead and avoid the additional lookup.

j0sh committed:
rpc: Networking v2.

j0sh committed:
rpc: Add generated go code.

j0sh committed:
rpc: Generate and fetch certs.

j0sh committed:
rpc: Implement broadcaster segment signing.

j0sh committed:
rpc: Integrate server with transcoder.

j0sh committed:
node: Integrate broadcaster with RPC client.

j0sh committed:
node: Watch for jobs in broadcaster.

j0sh committed:
stubclient: Cosmetics.

j0sh committed:
rpc: Get StreamIDs from orchestrator.

j0sh committed:
node: Propagate transcoder errors back to broadcaster.

j0sh committed:
rpc: Verify issuance and expiry of auth token.

j0sh committed:
cert: Generate fresh certs on each restart for now.

j0sh committed:
rpc: Propagate auth errors back to the broadcaster.

j0sh committed:
rpc: Verify job assignment in GetTranscoder request.

j0sh committed:
doc: Add networking spec.

yondonfu commented on doc/networking.md:
```diff
@@ -0,0 +1,132 @@
+
+Verification of `TranscoderRequest` consists of the following steps:
+1. Fetch the job based on JobId.
+    * TODO distinguish if an orchestrator is lagging behind and hasn't yet received the block with the job. Might require incorporating the node's **current** block number into the request. `CreationBlock`  as stored in the job only approximates when the job is submitted to the chain, not when it is included in a block and assigned to a transcoder.
```
The `job.creationBlock` stored in the JobsManager contract actually does reflect the block that the job transaction was included in a block and assigned to a transcoder. References to the current block (i.e. using the Solidity built in `block.number`) within a contract always refer to the block that a transaction is being included in (not the block at which the transaction is submitted).

j0sh replied:
Gotcha. Just curious, how are uncles handled?

yondonfu replied:
Uncles need to be handled by the client - transactions included in uncle blocks need to be included again in blocks on the main chain in order for them to be a part of the chain history. So it would be possible for an orchestrator to see a job transaction in a block, only to realize that the block is an uncle such that the job transaction does not yet exist on the main chain (but it could be included again in a subsequent block). We don't handle these types of cases right now, but we could implement a required number of confirmations to wait so clients can be more certain that a transaction was not included in an uncle (until we have finality with PoS one day so we don't need to wait for confirmations...)

yondonfu commented on doc/networking.md:
```diff
@@ -0,0 +1,132 @@
+
+Verification of `TranscoderRequest` consists of the following steps:
+1. Fetch the job based on JobId.
+    * TODO distinguish if an orchestrator is lagging behind and hasn't yet received the block with the job. Might require incorporating the node's **current** block number into the request. `CreationBlock`  as stored in the job only approximates when the job is submitted to the chain, not when it is included in a block and assigned to a transcoder.
```
Hm I wonder if the possibility of gaps in blocks between the broadcaster and orchestrator could introduce some potential annoyance for an orchestrator - a broadcaster could send `TranscoderRequest` with a block number that is a number of blocks in the future, basically stating that the head of the chain is N when it is in fact M (which is the orchestrator's latest synced block) such that M < N. The orchestrator does not have an easy way to determine if its latest synced block is the head of the chain so it might believe that it is just lagging behind and wait until block N to see if it can retrieve the job at that point. But, that job might not exist if the broadcaster lied about the existence of the job at block N leading the orchestrator to wait for N - M blocks for nothing. Perhaps if broadcasters include the creation block for a job in `TranscoderRequest` and the creation block is greater than the latest synced block of the orchestrator, the orchestrator could set some threshold of blocks that it is willing to wait or else it will just ignore the request?

j0sh replied:
Yeah -- in the split-brain case, the best thing to do would be to return an appropriate error code (eg, 404 "Job Not Found" or "Block Not Found"), and have the client be responsible for re-trying later. That way we don't tie up any resources on the server waiting for a block that may never come.

Also not sure if we really need the block number, or would the job ID be enough given enough time/retries?

I'm inclined to wait on this case and make it watertight after the merge, since the client doesn't yet do other corrective action as a result of server errors.

yondonfu replied:
I think returning an appropriate error if the orchestrator cannot find the job seems reasonable. And yeah I think if we're just returning an error and not trying to handle any block syncing disparities between the orchestrator and broadcaster in favor of having the broadcaster just retry, the job ID should be enough

yondonfu commented on doc/networking.md:
```diff
@@ -0,0 +1,132 @@
+Processing a `/segment` request consists of the following steps:
+
+1. Verify the credentials generated by the orchestrator
+2. Verify that the job hasn't expired.
```
Just thinking about the scenario where the orchestrator and transcoder are physically distinct entities - this step would require the transcoder to have access to the ETH network if it is attempting to be self-sufficient by verifying the job expiration itself via a contract lookup and by having knowledge of the current block number. But, if it does not need to be self-sufficient (the trusted case), the auth token it receives from the orchestrator could contain the approximate job end time as you mentioned so the transcoder wouldn't need any access to the ETH network. It might not be very exact given varying block times, but perhaps it could be decent estimate...

dob replied:
I think the idea of transcoders not needing to access the chain - and instead they just take the orchestrators directives - as appealing. But maybe I'm looking too far ahead to transcoder races, when right now we should just keep focus on this first implementation to get to a stable network.

j0sh replied:
Yeah, I punted on the "off-chain transcoder" issue for now, because there are still some unresolved questions, such as deposit checking when a broadcaster has multiple concurrent jobs.

yondonfu commented on doc/networking.md:
````diff
@@ -0,0 +1,132 @@
+* **Livepeer-Segment**
+Proves that the broadcaster generated this segment. Serialized protobuf struct, base64 encoded.
+```protobuf
+message SegData {
````
Is the broadcaster signature going to be included in this message or is it external to this message and stored in another header field?

j0sh replied:
Yes, the broadcaster's segment signature is in the `sig` field of SegData. In turn, that is protobuf-encoded and base64-encoded, and stored in the `Livepeer-Segment` header.

yondonfu replied:
Oops didn't see the `sig` field.

yondonfu commented on doc/networking.md:
```diff
@@ -0,0 +1,132 @@
+  string credentials = 3;
+
+  // Transcoded streamId list to update the master manifest on the broadcaster.
+  map<string, string> streamIds = 50;
```
This is might just be a protocol buffer thing - why is the field number 50 and not 4 (continuing from the previous field number of 3)?

j0sh replied:
In protobufs, field numbers 1-15 take one byte to encode, 16-2047 take two bytes, so the recommendation is to give less frequently occurring fields a higher number, and leave room in the lower end of the range for future expansion. (Why did I pick 50 specifically, then? No real reason other than I couldn't remember what the range was.) https://developers.google.com/protocol-buffers/docs/proto#assigning-field-numbers

In the case of StreamIds -- I wasn't sure if we would always be returning this type of data in future iterations of the protocol, since it's quite specific to how the relay protocol currently works. But perhaps that assumption is a bit premature here, and it's not really implemented consistently for the other structs either. Will see about cleaning it up.

dob commented on doc/networking.md:
```diff
@@ -0,0 +1,132 @@
+Taken from the `authType` field in TranscoderInfo.
+* **Credentials**
+Taken from the `credentials` field in TranscoderInfo.
+* **Livepeer-Segment**
```
Did you think about including this protocol specific metadata in the body of the request itself. I get that auth + credentials are usually included in headers, but the livepeer-segment piece seems kind of application/content specific. Ok with it either way, but usually I see APIs using headers to include request specific information, but the body containing all the app specific information.

j0sh replied:
The advantage of putting the livepeer-segment in the headers is that it allows us to verify the broadcaster signature without downloading the entire body of the request. Headers are just a few KB, while the segment is a few MB. In that sense, it's another authentication mechanism.

j0sh committed:
jobservice: Don't invoke the transcoder anymore.

With the new networking scheme, the job only needs to be claimed.

j0sh committed:
node: Check deposit prior to transcoding.

j0sh committed:
eth: Simplify job transcoder assignment retrieval.

In each case, the job was already being retrieved separately, so
avoid the duplicate retrieval in the AssigedTranscoder function.

yondonfu commented on core/livepeernode.go:
```diff
@@ -63,6 +63,13 @@ const (
+
+type SegChanData struct {
+	seg *SignedSegment
+	res chan error
```
Perhaps for clarity this should be named something like `errorChan` instead of `res` - at first glance, it might not be obvious for a reader that its a channel specifically for errors as opposed some non-error result value?

j0sh replied:
It's actually okay/expected to return a non-error (nil) value here since that means the transcoding succeeded. Still, I can make the change if you think it helps the clarity.

If Go was a language with proper sum types, `res` would be a `Maybe error` rather than a bare `error` type.

yondonfu replied:
Ah I see - I wasn't taking into account the usage of the `res` channel to block until the worker sends a signal that processing of a particular segment is complete. With that in mind I think keeping the name makes sense.

yondonfu commented on core/livepeernode.go:
```diff
@@ -125,32 +134,143 @@ func (n *LivepeerNode) Start(ctx context.Context, bootIDs, bootAddrs []string) e
+				n.claimMutex.Unlock()
+				return
+			case chanData := <-segChan:
+				chanData.res <- n.transcodeAndBroadcastSeg(&chanData.seg.Seg, chanData.seg.Sig, cm, tr, resultStrmIDs, broadcasters, config)
```
Maybe include a check `if err = n.transcodeAndBroadcastSeg(); err != nil {}` to avoid passing any non-error values down the channel to the consumer?

j0sh replied:
We actually need to return something in order to un-block the channel so the broadcaster gets a response. A nil is good; that means the transcoding succeeded. Otherwise, the error gets serialized into a string and sent back in the response. (Yeah, serializing the error is a bit rough for now, but it's a big improvement over old approach of "blindly send the segment up the network and hope it gets transcoded.")

j0sh committed:
readme: Update.

yondonfu commented on core/livepeernode.go:
```diff
@@ -125,32 +134,143 @@ func (n *LivepeerNode) Start(ctx context.Context, bootIDs, bootAddrs []string) e
+		PerformOnchainClaim: cm != nil,
+	}
+	go func() {
+		for {
```
Since this function starts an infinite for loop within a goroutine should it accept a cancellable context so that things here get cleaned up properly if for example the node exits while this loop is still running? The context below used for `transcodeAndBroadcastSeg()` timeouts could be a derived child of the passed in context

j0sh replied:
The loop isn't infinite; it times out after 10 minutes if there's no activity.

There's also the question of where any context should come from. The `ServeSegment` HTTP handler is what kicks off the transcoding loop, and the context contained within [Request](https://golang.org/pkg/net/http/#Request.Context) gets cancelled when the connection breaks, which is not what we want. Golang also discourages storing contexts, so we shouldn't really create a new one to be stored in the LivepeerNode for later cancellation.

Right now if the node exits (whether cleanly or not), there isn't really any lingering state that needs to be cleaned up within this function, so maybe punt on the context for now?

yondonfu replied:
Hm yeah that's tricky...I think its fine to punt on the context for now

yondonfu commented on eth/eventservices/jobservice.go:
```diff
@@ -204,7 +203,7 @@ func (s *JobService) doTranscode(job *lpTypes.Job) (bool, error) {
 func (s *JobService) RestartTranscoder() error {
 
-	eth.RecoverClaims(s.node.Eth, s.node.Ipfs, s.node.Database)
+	return eth.RecoverClaims(s.node.Eth, s.node.Ipfs, s.node.Database)
```
Should this be:

```
if err := eth.RecoverClaims(...); err != nil {
    return err
}
```

j0sh replied:
Ah nope -- we actually intentionally skip all the code past that point. Wanted to minimize the diff. This isn't the only place where we have dead code; we can do a clean-up pass after things have settled a bit. Added a comment as such.

yondonfu commented on server/rpc.go:
```diff
@@ -0,0 +1,462 @@
+	if orch.node == nil || orch.node.Eth == nil {
+		return nil
+	}
+	block, _ := orch.node.Eth.LatestBlockNum()
```
Check error returned here? If `block` is nil could get some nil pointer exceptions

j0sh replied:
Actually that's okay to ignore the error here because we return *big.Int only (and not an error) but none of the callers of `CurrentBlock()` actually check the result, so added that in...

The only caller was the `blockInRange` helper. Returned `true` if the block was nil; I figure it's better to leave the benefit of doubt here. If the transcoder has problems fetching the block, then it needs to get those problems fixed; we shouldn't compromise the broadcaster's job because of an unrelated, internal issue in checking validity. Totally open to discussion.

yondonfu replied:
Ah because `orch.node.Eth.LatestBlockNum()` uses a constant backoff with no max # of retries so in theory if it returns something it will always be a valid `*big.Int` (I actually don't think it should ever really return nil except for maybe the case where you're talking to a node with 0 blocks) or else it would just keep continuously retrying...perhaps later on the function return signature could be updated, don't think its necessary here though

> If the transcoder has problems fetching the block, then it needs to get those problems fixed; we shouldn't compromise the broadcaster's job because of an unrelated, internal issue in checking validity. 

I think that is fine for now, but in the future at a more stable point I think there is an argument for stopping everything you're doing if you are unable to perform a validity check properly

j0sh replied:
We subscribe to new block events within the JobService and write it to disk. This was done for job recovery, but we don't need to recover jobs anymore. The subscription could still be useful on its own though.

Rather than look up the block number via eth RPC every time, I was actually thinking it might be better to look up the block number from the database or even store it in a global var. Faster, more reliable, and less unnecessary load on the network. This is especially relevant here, since the lookup is being done during the transcoding flow, which is under tighter time constraints; don't want a geth stall to back up the transcoding.

yondonfu commented on server/rpc.go:
```diff
@@ -0,0 +1,462 @@
+			SeqNo:    uint64(segData.Seq),
+			Name:     "",
+			Data:     data,
+			Duration: 0.,
```
Do we need duration here?

j0sh replied:
Nope, and we don't need `Name` either. Removed both

yondonfu commented on eth/client.go:
```diff
@@ -887,6 +918,39 @@ func (c *client) ReplaceTransaction(tx *types.Transaction, method string, gasPri
 }
 
+func (c *client) WatchForJob(streamId string) (*lpTypes.Job, error) {
+	ctx, _ := context.WithTimeout(context.Background(), c.txTimeout)
```
`defer cancel()` here just for consistency around always cancelling contexts?

yondonfu commented on eth/eventservices/jobservice.go:
```diff
@@ -231,6 +230,48 @@ func (s *JobService) RestartTranscoder() error {
+	}
+	firstClaimBlock := new(big.Int).Add(job.CreationBlock, eth.BlocksUntilFirstClaimDeadline)
+	headersCh := make(chan *types.Header)
+	s.eventMonitor.SubscribeNewBlock(context.Background(), fmt.Sprintf("FirstClaimForJob%v", job.JobId), headersCh, func(h *types.Header) (bool, error) {
```
Perhaps we should make sure to call `Unsubscribe()` on the returned subscription (believe this was previously missing as well)

j0sh replied:
Sure -- added for this one, left the old one alone since it's dead code now.

yondonfu commented on eth/eventservices/jobservice.go:
```diff
@@ -231,6 +230,48 @@ func (s *JobService) RestartTranscoder() error {
 	return nil
 }
 
+func (s *JobService) claimJob(job *lpTypes.Job) {
```
Since this function is for monitoring whether the node should submit the first claim for a job maybe it should be named `submitFirstClaimForJob` or something similar

j0sh replied:
Sure, renamed to `firstClaim` , hope the brevity is OK.

ericxtang commented on eth/client.go:
```diff
@@ -887,6 +918,32 @@ func (c *client) ReplaceTransaction(tx *types.Transaction, method string, gasPri
+	ctx, _ := context.WithTimeout(context.Background(), c.txTimeout)
+	for {
+		sink := make(chan *contracts.JobsManagerNewJob)
+		sub, err := c.JobsManagerSession.Contract.JobsManagerFilterer.WatchNewJob(nil, sink, []ethcommon.Address{c.Account().Address})
```
I think this might run into the same issue we did for Infura in `eventmonitor.go`, and ended up having to add the retry logic (anywhere we used `backoff.Retry`)

j0sh replied:
Sure. Note that this already runs in a for-loop and retries on subscription errors, but not context timeouts. I can change it to cover both cases within`backoff.Retry` to be more consistent with the rest of the code.

One thing I dislike about `backoff.Retry` is that the API is clunky due to golang. In order to extract a non-error value from `Retry`, it forces us into the pattern of: declare an externally scoped variable, runtime create a closure for the var, explicitly mutate the var, the finally return that after error checking. With polymorphic HOFs, we could replace the `for { }` with a `func` and be done with it.

j0sh replied:
Okay, the changes are in. Note that the watcher API has changed since the review -- now we do the job lookup inside this watcher and return the `Job` directly. I've opted not to retry the `GetJob` call if that fails, and instead we reconstruct the job from the event data, with the caveat that some fields may be missing. Whether any of that is a good idea, is open to discussion.

https://github.com/livepeer/go-livepeer/blob/948fe4b2971ddacd485f0ace392f101fc9091af8/eth/client.go#L953-L973

yondonfu commented on server/cert.go:
```diff
@@ -0,0 +1,107 @@
+	tmpl := x509.Certificate{
+		SerialNumber:          serial,
+		NotBefore:             time.Now(),
+		NotAfter:              time.Now().Add(8765 * time.Hour), // XXX fix fix fix
```
Should we define a constant for the default validity duration?

yondonfu commented on server/cert.go:
```diff
@@ -0,0 +1,107 @@
+		SerialNumber:          serial,
+		NotBefore:             time.Now(),
+		NotAfter:              time.Now().Add(8765 * time.Hour), // XXX fix fix fix
+		KeyUsage:              x509.KeyUsageCertSign,
```
Does the template certificate need to define the `SignatureAlgorithm` field (https://golang.org/pkg/crypto/x509/#SignatureAlgorithm)?

j0sh replied:
Testing seems to be OK without it, and in any case we ignore checking the signature for TLS, so...

Do the docs indicate that the field is required? I can't find anything saying so. From reviewing the implementation, it seems that a default value is chosen anyway based on the type of the cert. https://golang.org/src/crypto/x509/x509.go, `signingParamsForPublicKey`

yondonfu replied:
Was just asking because I didn't see anything in the docs and thinking about the case where we do check the signature, but based upon the implementation `signingParamsForPublicKey()`, it seems like its not required so we are 👍

yondonfu commented on server/mediaserver.go:
```diff
@@ -119,11 +120,38 @@ func (s *LivepeerServer) StartMediaServer(ctx context.Context, maxPricePerSegmen
+				}
+			}
+			uri, err := url.Parse(serviceURI) // should be publicly accessible
+			if err == nil && uri.Port() != "" {
```
Should we also stop here if `err != nil` in the case that the service URI returned from the registry was somehow malformed?

j0sh replied:
Sure, we don't want to enforce that right now though, to give existing transcoders time to adopt the service URI. I can add an empty check.

yondonfu commented on server/mediaserver.go:
```diff
@@ -274,6 +303,10 @@ func gotRTMPStreamHandler(s *LivepeerServer) func(url *url.URL, rtmpStrm stream.
 					}
 				}
+
+				if rpcBcast != nil {
```
It looks like `rpcBcast` won't be initialized until the new job is confirmed on-chain (https://github.com/livepeer/go-livepeer/blob/de1836eec5674ed7da481b9de2f237a86dca7405/server/mediaserver.go#L386) - would it make sense to drop segments and not run any of the above processing steps until `rpcBcast != nil`?

j0sh replied:
Most of the earlier steps are still run during the old `Broadcast` flow to allow subscribers to view the stream, so it's still useful even while the job is unconfirmed (or viewing locally / off-chain).

yondonfu commented on server/rpc.go:
```diff
@@ -0,0 +1,462 @@
+		glog.Error("Could not get job ", err)
+		return nil, fmt.Errorf("Missing job")
+	}
+	blk := orch.CurrentBlock()
```
Use the `blockInRange()` helper here?

yondonfu commented on server/rpc.go:
```diff
@@ -0,0 +1,462 @@
+	}
+
+	glog.Info("Listening for RPC on ", bind)
+	http.ListenAndServeTLS(bind, cert, key, &lp)
```
I don't think the `http` package level convenience functions like `ListenAndServeTLS` by default enforce a read/write timeouts from clients - perhaps we should consider configuring default timeout values for the HTTP server to drop slow clients if they take forever to read/write?

j0sh replied:
Sure, set it to 4 seconds for now, for the client and server code. If a RTT takes longer than that, segments are guaranteed to be backed up.

yondonfu reviewed (changes requested):
Just to clarify for personal understanding: the current certificate generation code has no relation with an orchestrator's ETH keys since every time the transcode server is started a new fresh key will be generated along with a fresh certificate - once we decide on a strategy for https://github.com/livepeer/go-livepeer/issues/437 we could use the orchestrator's ETH key for the root certificate and the orchestrator could optionally create new certificates whenever it wants that are signed by the root certificate. However, right now the orchestrator just sets up a fresh certificate + key so it can create a TLS enabled server - the broadcaster forms a TLS connection with the server, but skips the verification step. Is this an accurate understanding?

j0sh commented:
@yondonfu Your description of the TLS / cert behavior is correct.

@ericxtang @yondonfu Most of the suggested changes should be in by now.

j0sh committed:
claimmanager: Fix crash when claiming an empty unclaimed set.

Happens if we do a claim in order to satisfy the 255 block limit,
and the transcode loop subsequently expires without more claims.

yondonfu commented on README.md:
```diff
@@ -55,16 +55,15 @@ You can also build the executables from scratch.
 
-To broadcast, run `./livepeer_cli` and pick 'Broadcast Video'.  
-  * You should see your webcam becoming active and a manifestID printed on the screen.
+For full details, read out [Broadcasting guide](http://livepeer.readthedocs.io/en/latest/broadcasting.html).
```
"read the" or "read our"?

yondonfu commented on eth/client.go:
```diff
@@ -938,6 +935,63 @@ func (c *client) ReplaceTransaction(tx *types.Transaction, method string, gasPri
+					// maybe perform/retry the job lookup outside this loop
+					// but a retry may be unlikely to succeed, depending on the error
+					// Manually create the job for now.
+					// May have important fields missing!!!
```
IMO if `WatchForJob` is expected to be able to be reused by anyone then we should always return a job with all the fields populated (retrieved using a RPC call) and if it is expected to be only used by a broadcaster then we should always return a job partially filled with broadcaster related fields (i.e. reconstructed from the fields in the event - so in this case we could skip the RPC call completely). I don't think it is a problem right now since `WatchForJob` is used exclusively in `CreateTranscodeJob` which is only relevant for a broadcaster, but I think it could be confusing for usage later on if the function ends up being used elsewhere and it sometimes returns a job with all the fields and sometimes returns a job with only some of the fields

j0sh replied:
We could also refactor the `JobService` to use `WatchForJob`  rather than the old eventmonitor (which would also not expect have all fields populated) but agreed that there's potential to 'surprise' API users later on. We can add a note to the function warning future users about the potential discrepancy.

j0sh replied:
Note added:

https://github.com/livepeer/go-livepeer/blob/35f32fd32e3a03d3a598c57fe7f71fed8a3c49e1/eth/client.go#L938-L941

yondonfu reviewed (commented):
Changes look good to me. Only additional comment that might warrant a change is about consistency with fields included in job returned by `WatchForJob`

yondonfu reviewed (approved):
LGTM

j0sh commented:
Added a fix to the claimmanager for a crash that was happening under certain conditions with the new code: https://github.com/livepeer/go-livepeer/pull/447/commits/52c2382bcb94693248287ba7a9807775832a9643

ericxtang reviewed (approved):
🚢

j0sh commented:
Updated API doc; removed incorrect note about the `CreationBlock` reflecting the block it was created at.

j0sh commented:
Changed the content type for uploads to `video/MP2T` to reflect the actual mime type of the file that we're uploading. More standard, and allows for differentiation when we support different formats.

j0sh committed:
livepeer: Use websockets for non-transcoders.

Broadcasters now need to watch for jobs, and watching doesn't
work over plain HTTP RPC.

j0sh committed:
mediaserver: Remove the video nonce unconditionally.

We add a nonce even if metrics monitoring is not enabled, so
this prevents a leak.

j0sh commented:
Few more changes as a result of Rinkeby testing. Most were small typos that were fixed up, but there are a couple behavioral changes that are being kept as separate commits:

* Broadcasters now need to use Websocket RPC rather than HTTP RPC. https://github.com/livepeer/go-livepeer/pull/447/commits/16ba6407bc2e9e4bd154ff3c32ddef4bcc1f836d

* Small bugfix around nonces https://github.com/livepeer/go-livepeer/pull/447/commits/6101359b4c8d74fdfa58fc53845c3168f63c40d4

* Increased the request timeout since too many transcoding segments were timing out. This is something of a concern, but note that if the initial upload takes a consistent amount of time, and transcodes complete under the 4 second limit, then it should be OK for the most part, given that distribution afterwards is also consistent.

Also added grpc connection retries to address synchronization issues with the orchestrator. Inadvertedly wrote this against the branch for [restartable broadcasts](https://github.com/livepeer/go-livepeer/pull/428). There were significant changes in the same area for restartable broadcasts, so left the commit in the restartable branch, rather than attempt to back-merge into this branch. https://github.com/livepeer/go-livepeer/pull/428/commits/69f99a491eb63ed8405a476852af84c2573b37b1

j0sh committed:
rpc: Return 200OK early after checking the upload.

Clients can use this as an indication that the upload has
completed, and further approximate the transcode time from there.

j0sh commented:
Added a small enhancement that was useful during this week's testing: https://github.com/livepeer/go-livepeer/pull/447/commits/7e70b5c60699155a9deefa1ebc0e8b3b9703e763

The 200OK response header is returned early after the upload. This allows the broadcaster to distinguish how long the upload took, and how long the transcode took. Here's an example script that I use to tail the broadcaster's logs to get some per-segment statistics: https://gist.github.com/j0sh/ea6624d55e519d5146e6568637582140

j0sh committed:
rpc: Reduce back pressure.

Force HTTP/2 : This was a bug all along, but helps nonetheless
by eliminating some head-of-line blocking with HTTP/1.1, which
prevented some interleaving of segment processing.

Make the segment submission asynchronous within a goroutine:
Increases network utilization and decreases the elpased time
between segments, which would otherwise be increasing each time
the upload + transcode took more than 4 seconds. This would
sometimes result in 6 or 7 seconds between segments.

Async submission isn't entirely correct; for example, during
large backlogs, scheduling could cause segments to be delivered
out of order at the transcoder. The most correct thing to do is
to ensure that only one segment is in-flight on the wire at any
given time. However, this quick and easy approach is OK for now.

j0sh committed:
claimmanager: Avoid storing video data in memory.

This relieves memory build-up when managing long segment sequences.

Note that we still retain per-segment data in memory until claims
is complete, and this could eventually exhaust available RAM, but
this change should give us much more runway.

j0sh commented:
Some further changes to relieve back pressure during the transcoding process: https://github.com/livepeer/go-livepeer/pull/447/commits/690d41eef081afb7bb2c21f1ffb8786cfae6a1e1

Force HTTP/2 : This was a bug all along, but helps nonetheless
by eliminating some head-of-line blocking with HTTP/1.1, which
prevented some interleaving of segment processing.
(Underlying TCP connections were still being reused.)

Make the segment submission asynchronous within a goroutine:
Increases network utilization and decreases the elpased time
between segments, which would otherwise be increasing each time
the upload + transcode took more than 4 seconds. This would
sometimes result in 6 or 7 seconds between segments.

Async submission isn't entirely correct; for example, during
large backlogs, scheduling could cause segments to be delivered
out of order at the transcoder. The most correct thing to do is
to ensure that only one segment is in-flight on the wire at any
given time. However, this quick and easy approach is OK for now.

Note that if the end-to-end segment processing (upload, transcode, etc)
consistently takes 6 or 7 seconds, a backlog builds up and it eventually
'collapses' which times out the next several segments (sometimes as
many as a few dozen, depending on how long it takes for the network
to recover to a stable state).

From testing, the culprit has usually been the network between my
local machine and the remote machine. RTMP ingest to a cloud
broadcaster helps a lot, since home connections (at least, *my*
home connection) doesn't handle the bursty traffic very well.

j0sh commented:
Attempted to fix some issues we had with looking up a transcoder here: https://github.com/livepeer/go-livepeer/pull/478/commits/2d1d5301568d15b8a560456e09b7183dbeb72d03 

Some notes about the implementation. We can't really rely on recreating the job from the event data as a fallback, because it's missing some important fields: `EndBlock` (needed for DB) and `CreationRound` (needed to elect the active transcoder). In the medium term, it'd be nice to add those two fields to the job created event, but that's a protocol change.

We could approximate both fields locally rather than retry the RPC call, but I think we'd rather sacrifice latency for correctness in this case. This is one reason that a [separate job creation step](https://github.com/livepeer/go-livepeer/issues/429) would be nice, which would be implemented alongside [multi-stream support](https://github.com/livepeer/go-livepeer/issues/481).

Hopefully this fixes  https://github.com/livepeer/go-livepeer/issues/495 and https://github.com/livepeer/go-livepeer/issues/493. The fix was done half blind. When creating jobs in a loop for a few hours, `eth.AssignedTranscoder` fails occasionally (and has recovered with this fix), but I haven't seen `eth.GetJob` return an empty job yet. If we still get empty jobs even after this, let me know.

cc @ericxtang @yondonfu
